File: maximum-good-people-based-on-statements-2151/Solution.java

/**
 * LeetCode problem # 2151, 'Maximum Good People Based on Statements'
 * https://leetcode.com/problems/maximum-good-people-based-on-statements/
 */

import java.util.*;

class Solution {

    private boolean isValid(boolean[] vis, int[][] statements) {
        int n = statements.length;
        for (int i = 0; i < n; ++i) {
            if (vis[i]) {
                int[] curr = statements[i];
                for (int j = 0; j < n; ++j) {
                    if (curr[j] == 2) {
                        continue;
                    }
                    if (curr[j] == 0 && vis[j]) {
                        return false;
                    }
                    if (curr[j] == 1 && !vis[j]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public int maximumGood(int[][] statements) {
        int n = statements.length;
        int res = 0;

        // generate the 2^n integers (2 <= n <= 15 so ok)
        for (int i = 0; i < (1 << n); ++i) {
            boolean[] vis = new boolean[n];

            int k = 0;
            int temp = i;
            while (temp != 0) {
                if (temp % 2 == 1) {
                    vis[k] = true;
                }
                temp >>= 1;
                ++k;
            }
            // System.out.println(Arrays.toString(vis));
            if (isValid(vis, statements)) {
                // # of set bits in the binary representation of i
                // (good people in current combo)
                res = Math.max(res, Integer.bitCount(i));
            }
        }
        return res;
    }
}

File: maximum-good-people-based-on-statements-2151/solution.js
/**
 * LeetCode problem # 2151, 'Maximum Good People Based on Statements'
 * https://leetcode.com/problems/maximum-good-people-based-on-statements/
 */

/**
 * @param {number[][]} statements
 * @return {number}
 */
function maximumGood(statements) {
  const n = statements.length;
  let max = 0;

  function generateAll(arr, index, count) {
    if (index == arr.length) {
      let res = check(arr);
      if (res) {
        max = Math.max(max, count);
      }
      return;
    }
    generateAll(arr, index + 1, count);
    arr[index] = 1;

    generateAll(arr, index + 1, count + 1);
    arr[index] = 0;
  }

  function check(arr) {
    for (let i = 0; i < n; ++i) {
      if (arr[i] == 0) {
        continue;
      }
      for (let j = 0; j < n; ++j) {
        if (statements[i][j] == 2) {
          continue;
        }
        if (
          (statements[i][j] == 1 && arr[j] == 0) ||
          (statements[i][j] == 0 && arr[j] == 1)
        ) {
          return false;
        }
      }
    }
    return true;
  }

  const arr = Array(n).fill(0);
  generateAll(arr, 0, 0);
  return max;
}
